### PR TITLE
[Feature] Contribution management

### LINKED ISSUE #780: [Feature] Contribution management
We must have the ability to edit a contribution on a plan. See associatedemployers/igneous#334 for contribution field requirement.

### PR DESCRIPTION
**Pull request by JavaLavaMT**
closes #780 
close with associatedemployers/igneous#340

Screen Shot 2020-08-31 at 1 21 42 PM
Screen Shot 2020-08-31 at 1 21 55 PM
Screen Shot 2020-08-31 at 1 22 02 PM
Screen Shot 2020-08-31 at 1 22 09 PM
Screen Shot 2020-08-31 at 1 22 18 PM
Screen Shot 2020-08-31 at 1 22 29 PM
Screen Shot 2020-08-31 at 1 23 10 PM

### CODE REVIEW

File: app/components/benefits/contributions-row.hbs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            </div>
+            <div class="item" data-value="percent of cost">
+              %
+            </div>
+          </div>
+        </UiDropdown>
+      </div>
+      {{#if this.wellnessPlan}}
+        <div class="field">
+          <label for="family-Wellness-Modifier">On Wellness Plan</label>
```
Comment (James1x0) [inline app/components/benefits/contributions-row.hbs:131]:
```suggestion
          <label for="family-Wellness-Modifier">On Wellness Plan</label>
```

Class names and ids should always be lower case

File: app/components/benefits/contributions-table.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+
+export default class PlanContributionsTableComponent extends Component {
+  contributionGroups = [ 'Employee', 'Employee & Spouse', 'Employee & Children',' Family' ]
```
Comment (James1x0) [inline app/components/benefits/contributions-table.js:5]: I think there's a rogue space in one of your strings here.

File: app/components/benefits/contributions-table.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+  @action
+  cancel () {
+    this.args.cancel();
```
Comment (James1x0) [inline app/components/benefits/contributions-table.js:9]: I believe both cancel and save here can just be removed as intermediaries. In the template you can write:

{{fn @save  this.wp}}

File: app/components/benefits/plan.hbs [hunk 4]
```diff
@@ -28,25 +28,25 @@
       {{#if @plan.ratesEmployee}}
         <div class="line-item-value">
           <i>Employee</i>
-          <u>{{format-money @plan.ratesEmployee}}</u>
+          <u>{{format-money @plan.ratesEmployee}}{{if @plan.contributionsEmployeeWellnessModifier (concat " (" (format-money @plan.contributionsEmployeeWellnessModifier) "w/ wellness)") }}</u>
```
Comment (James1x0) [inline app/components/benefits/plan.hbs:31]: Unnecessary complication, don't use concat. You're in a template, just write it out in a block {{#if}}

File: app/components/benefits/plan.hbs [hunk 5]
```diff
@@ -62,6 +62,54 @@
       {{/if}}
     </div>
   </div>
+  <div class="extra content plan-card__contributions">
```
Comment (James1x0) [inline app/components/benefits/plan.hbs:65]: BEM 💪 🤘

File: app/components/benefits/plan.hbs [hunk 6]
```diff
@@ -62,6 +62,54 @@
...
+            <u>{{format-money @plan.contributionsSpouseAmount}}</u>
+          {{else}}
+            <u>{{@plan.contributionsSpouseAmount}}%</u>
+          {{/if}}
+        </div>
+      {{/if}}
+      {{#if @plan.contributionsChildrenAmount}}
+        <div class="line-item-value">
+          <i>Children</i>
+          {{#if (eq @plan.contributionsChildrenType 'dollar')}}
```
Comment (James1x0) [inline app/components/benefits/plan.hbs:94]: I'm sure that this is a template lint error :/
You should use double-quotes in templates.

File: app/controllers/account/benefits/plans/contributions/edit-all.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+      await this.data.saveRecord(plans[i]);
+    }
+    this.transitionToRoute('account.benefits.plans');
+  }
+
+  @action
+  cancel () {
+    this.model.forEach(plan => {
+      plan.rollbackAttributes();
+    });
```
Comment (James1x0) [inline app/controllers/account/benefits/plans/contributions/edit-all.js:22]: While you are working on this, if we have a record array, like `model`, we can actually make use of Ember's ArrayProxy method called `invoke`

This becomes:
```
this.model.invoke('rollbackAttributes');
```

File: app/components/benefits/contributions-table.hbs [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+      <Benefits::ContributionsRow @plan={{@plan}} />
+    {{/each}}
+    </tbody>
+  </table>
+  <div class="ui divider"></div>
+    <Form::ActionButton
+      type="submit"
+      class="ui large fluid green button">
+      Complete
+    </Form::ActionButton>
```
Comment (James1x0) [inline app/components/benefits/contributions-table.hbs:25]: I'm somewhat expecting you to make use of a "working" or "loading" state here. Too often, people will click a save button 3 times and cause errors.

File: app/serializers/plan.js [hunk 9]
```diff
@@ -32,6 +33,36 @@ export default class Employee extends ApplicationSerializer {
...
+        wellnessModifier: json.contributionsChildrenWellnessModifier,
+        type:             json.contributionsChildrenType
+      },
+      family: {
+        amount:           json.contributionsFamilyAmount,
+        wellnessModifier: json.contributionsFamilyWellnessModifier,
+        type:             json.contributionsFamilyType
+      }
+    };
+
```
Comment (James1x0) [inline app/serializers/plan.js:65]: Sorta confused why you didn't use your existing methodology here: https://github.com/associatedemployers/granite/pull/799/files#diff-8c20260321c10644ee652fca0edfd7d5R27-R36

File: app/templates/account/settings/index/features.hbs [hunk 10]
```diff
@@ -16,6 +16,19 @@
     <UiCheckbox @class="toggle" @label="On" @checked={{this.model.collectEEO}} @onChange={{pipe (set this.model.collectEEO _) (fn this.data.saveRecord this.model)}} />
   </div>
 
+  {{#if this.model.benefitsEnabled}}
+    <div class="field">
+      <label for="wellnessPlan">
```
Comment (James1x0) [inline app/templates/account/settings/index/features.hbs:21]:
```suggestion
      <label for="wellness-plan">
```

File: app/templates/account/settings/index/features.hbs [hunk 11]
```diff
@@ -16,6 +16,19 @@
...
+        Wellness Plan<br />
+        <small class="light text">
+        Does you company offer a wellness plan?, Enable this to change the contribution for wellness participants.
```
Comment (James1x0) [inline app/templates/account/settings/index/features.hbs:24]:
```suggestion
          Does your company offer a wellness plan? Enable this to change benefit contributions for wellness participants. 
```

File: app/templates/account/settings/index/features.hbs [hunk 12]
```diff
@@ -16,6 +16,19 @@
...
+  {{#if this.model.benefitsEnabled}}
+    <div class="field">
+      <label for="wellnessPlan">
+        Wellness Plan<br />
+        <small class="light text">
+        Does you company offer a wellness plan?, Enable this to change the contribution for wellness participants. 
+        </small>
+      </label>
+      Off
+      <UiCheckbox @class="toggle" @label="On" @checked={{this.model.wellnessPlan}} @onChange={{pipe (set this.model.wellnessPlan _) (fn this.data.saveRecord this.model)}} />
```
Comment (James1x0) [inline app/templates/account/settings/index/features.hbs:28]:
```suggestion
      <UiCheckbox @id="wellness-plan" @class="toggle" @label="On" @checked={{this.model.wellnessPlan}} @onChange={{pipe (set this.model.wellnessPlan _) (fn this.data.saveRecord this.model)}} />
```

File: app/components/benefits/contributions-table.js (hunk 2 above)
Comment (JavaLavaMT) [reply app/components/benefits/contributions-table.js:5]: 👁 found it

File: app/components/benefits/contributions-table.js (hunk 3 above)
Comment (JavaLavaMT) [reply app/components/benefits/contributions-table.js:9]: I like that a lot better! good call

File: app/components/benefits/contributions-row.hbs (hunk 1 above)
Comment (JavaLavaMT) [reply app/components/benefits/contributions-row.hbs:131]: 🏁

File: app/components/benefits/plan.hbs (hunk 6 above)
Comment (JavaLavaMT) [reply app/components/benefits/plan.hbs:94]: 🏁

File: app/controllers/account/benefits/plans/contributions/edit-all.js (hunk 7 above)
Comment (JavaLavaMT) [reply app/controllers/account/benefits/plans/contributions/edit-all.js:22]: learned something new

File: app/serializers/plan.js (hunk 9 above)
Comment (JavaLavaMT) [reply app/serializers/plan.js:65]: @James1x0 

I had problems with the dynamic "titles"
i.e. 
`json.contributions['$title']`

with the employee, it was a 
json['$title']

File: app/components/benefits/contributions-table.hbs (hunk 8 above)
Comment (JavaLavaMT) [reply app/components/benefits/contributions-table.hbs:25]: I totally was too, just completely forgot

File: app/components/benefits/plan.hbs (hunk 4 above)
Comment (JavaLavaMT) [reply app/components/benefits/plan.hbs:31]: 🏁

File: app/templates/account/employee/index/edit/job.hbs [hunk 13]
```diff
@@ -130,3 +130,105 @@
     </div>
   </UiDropdown>
 </div>
+{{#if this.company.benefitsEnabled}}
+  <h3 id="benefit-contributions" class="ui horizontal header divider section">
+    <i class="heartbeat icon"></i>Benefit Contributions
+  </h3>
+  {{#if this.company.wellnessPlan}}
+    <UiCheckbox @class="checkbox" @label="Wellness Plan" @checked={{readonly this.model.wellnessPlan}} @onChange={{set this.model.wellnessPlan _}} />
```
Comment (James1x0) [inline app/templates/account/employee/index/edit/job.hbs:138]:
```suggestion
    <UiCheckbox @class="checkbox" @label="Employee is on wellness plan" @checked={{readonly this.model.wellnessPlan}} @onChange={{set this.model.wellnessPlan _}} />
```

File: app/templates/components/benefits/plan.hbs [hunk 14]
```diff
@@ -54,7 +54,7 @@
           <i>Fixed</i>
           <u>
             {{format-money @plan.ratesFixed}}
-            {{#if (and @plan.lifeCoverage (or (eq @plan.type 'L') (eq @plan.type 'DL')))}}
+            {{#if (and @plan.lifeCoverage (or (eq @plan.type "L") (eq @plan.type "DL")))}}
```
Comment (James1x0) [inline app/templates/components/benefits/plan.hbs:57]:
```suggestion
            {{#if (and @plan.lifeCoverage (eq @plan.type "L"))}}
```

File: app/components/benefits/plan.hbs [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+  <div class="extra content plan-card__rates">
+    <strong>Rates</strong>
+    <div class="ui list">
+      {{#if @plan.ratesEmployee}}
+        <div class="line-item-value">
+          <i>Employee</i>
+          <u>
+            {{format-money @plan.ratesEmployee}}
+            {{#if @plan.contributionsEmployeeWellnessModifier}}
+              {{format-money @plan.contributionsEmployeeWellnessModifier}} w/ wellness
```
Comment (James1x0) [inline app/components/benefits/plan.hbs:34]: I really don't like how this is displayed. It seems to break the line flow. Let's do this, make this appear as:

$100.20/**$99.20**

with the second value having a tooltip titled, "With wellness modifier of -$1.00"

**Review by James1x0 — CHANGES_REQUESTED** [review]
There's a few things that still need adjustment

File: app/components/benefits/contributions-row.hbs [hunk 16]
```diff
@@ -0,0 +1,140 @@
+  <tr>
+    <td data-label="plan">
+      <h4>{{@plan.label}}</h4>
+      <div>
+      {{@plan.name}}
```
Comment (James1x0) [inline app/components/benefits/contributions-row.hbs:5]:
```suggestion
        {{@plan.name}}
```

File: app/components/benefits/contributions-row.hbs [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+            </div>
+            <div class="item" data-value="percent of cost">
+              %
+            </div>
+          </div>
+        </UiDropdown>
+      </div>
+      {{#if this.wellnessPlan}}
+        <div class="field">
+          <label for="family-Wellness-Modifier">On Wellness Plan</label>
```
Comment (James1x0) [inline app/components/benefits/contributions-row.hbs:131]:
```suggestion
          <label for="family-wellness-modifier">On Wellness Plan</label>
```
**Always** lowercase classes and ids.

File: app/components/benefits/contributions-row.hbs [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+          <div class="ui left icon input pay-rate-number">
+            <EmberTooltip>This is a rate adjustment for an employee on the wellness plan. This should be a negative number"</EmberTooltip>
+            <Input id="family-Wellness-Modifier" @type="number" @max={{0}} @value={{mut @plan.contributionsFamilyWellnessModifier}} />
```
Comment (James1x0) [inline app/components/benefits/contributions-row.hbs:134]:
```suggestion
            <Input id="family-wellness-modifier" @type="number" @max={{0}} @value={{mut @plan.contributionsFamilyWellnessModifier}} />
```
**Always** lowercase classes and ids.

File: app/components/benefits/plan.hbs [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+      {{@plan.label}} <i class="ui {{@plan.icon}} icon"></i>
+    </div>
+    <div class="header">{{@plan.name}}</div>
+    <div class="description plan-card__description">{{@plan.description}}</div>
+  </div>
+  {{#if @plan.networkName}}
+    <div class="extra content plan-card__network">
+      <strong>Network</strong>
+      <div class="basic segment">
+            <div class="line-item-value">
```
Comment (James1x0) [inline app/components/benefits/plan.hbs:13]: Indentation here is really wonky

File: app/components/benefits/plan.hbs [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+              <i>{{@plan.networkName}}</i>
+              <u>
+                <a href="" {{on "click" (prevent-default (toggle "showDescription" this))}}>{{if this.showDescription "Show Less" "Show More"}}</a>
```
Comment (James1x0) [inline app/components/benefits/plan.hbs:16]:
```suggestion
                <a href="#" {{on "click" (prevent-default (toggle "showDescription" this))}}>{{if this.showDescription "Show Less" "Show More"}}</a>
```

File: app/templates/account/benefits.hbs [hunk 21]
```diff
@@ -20,6 +20,7 @@
             <a href="#" id="action-item__action--delay" class="item" {{!--{{on "click" add}} --}}>
               <i class="ellipsis horizontal icon"></i>Manually add a plan
             </a>
+            <LinkTo @route="account.benefits.plans.contributions.edit-all" class="item"><i class="edit icon"></i>Edit Contributions</LinkTo>s
```
Comment (James1x0) [inline app/templates/account/benefits.hbs:23]:
```suggestion
            <LinkTo @route="account.benefits.plans.contributions.edit-all" class="item"><i class="edit icon"></i>Edit contributions</LinkTo>
```

**Comment by JavaLavaMT** [comment]
image

image

**Comment by JavaLavaMT** [comment]
fixed $ 
image

**Review by James1x0 — APPROVED** [review]

**Pull request merged**